File: proxy/websocket.go
package main

import (
    "fmt"
    "time"
	"encoding/json"
	"sync"
	"sync/atomic"
	"net/http"
	"log"

	"github.com/gorilla/websocket"
)

// This serves as our middle man to handle multiple concurrent websocket connextions for our applications
// Any one client can request from another connected app, rather than being limited to point to point
// This avoids a P2P mesh where every app needs to make its own WS connection to each other app (N^2)
var (
	// Track WebSocket connections and mutex for safe concurrent access
    wsConns = make(map[string]*websocket.Conn)
    wsMu sync.Mutex

	// Track the state of outbound websocket requests to clients
	id_counter uint64
	pendingRequests	sync.Map

    upgrader = websocket.Upgrader{
        ReadBufferSize:  1024,
        WriteBufferSize: 1024,
        CheckOrigin:     func(r *http.Request) bool { return true },
    }
)

// Normalized WebSocket message structure to maintain a consistent format for all events sent across the WebSocket bridge
type WSEnvelope struct {
	Event     string          `json:"event"`     			// e.g., "login", "packet_send", "packet_receive"
	Target    string          `json:"target"`    			// e.g., "backend", "script_api", "discord"
	ID        string          `json:"id,omitempty"` 		// ID of the request, used for callbacks if needed
	Timestamp int64           `json:"timestamp"` 			// Unix timestamp tracking when the event was received
	Payload   json.RawMessage `json:"payload"`   			// Keeps data raw/unparsed until it reaches its destination
}

func listenToClient(clientType string, conn *websocket.Conn) {
	// Close the WebSocket registry and close the connection when the client disconnects
	defer func() {
		log.Printf("Websocket client was disconnected: \"%s\"", clientType)
		wsMu.Lock()
		delete(wsConns, clientType)
		wsMu.Unlock()
		conn.Close()
	}()

	for {
		// Read incoming messages from this WebSocket connection
		_, message, err := conn.ReadMessage()
		if err != nil {
			break
		}
		// Parse the envelope to see where this message is supposed to go
		var envelope WSEnvelope
		if err := json.Unmarshal(message, &envelope); err != nil {
			continue
		}

		// Intercept and check if this fulfills an outbound request the proxy made
		if ch, found := pendingRequests.Load(envelope.ID); found {
			ch.(chan json.RawMessage) <- envelope.Payload
			continue 
		}

		// Check if this request is meant to be handled internally instead of rerouted
		if ignore := handleInternalRequest(&envelope); ignore {
            continue
        }

		// Forward the raw message to the target client if they are online
		wsMu.Lock()
		targetConn, online := wsConns[envelope.Target]
		wsMu.Unlock()

		if online {
			// Handoff the sender back to the target where it can then know who to respond back to. Example:

			// 1. Discord bot sends:   		{ "event": "get_players",			"target": "script_api" }
			// 2. Go Proxy receives:		{ "event": "get_players", 	  		"target": "script_api" }
			// 3. Go rewrites target: 		payload.Target = clientType = "discord"
			// 4. Forward to script:   		{ "event": "get_players",     		"target": "discord" }
			// 5. Script API receives:  	{ "event": "get_players",       	"target": "discord" }
			// 6. Script API responds:  	{ "event": "get_players_response", 	"target": "discord", "payload": { ... } }
			// 7. Go Proxy receives:    	{ "event": "get_players_response", 	"target": "discord", "payload": { ... } }
			// 8. Go rewrites target:   	payload.Target = clientType = "script_api"
			// 9. Forward back home:		{ "event": "get_players_response", 	"target": "script_api", "payload": { ... } }
			// 10. Discord bot receives: 	{ "event": "get_players_response", 	"target": "script_api", "payload": { ... } }

			// Now the original request knows it got its response back completing a round trip (using a request id)
			envelope.Target = clientType
			envelope.Timestamp = time.Now().UnixMilli()

            fullMessage, err := json.Marshal(envelope)
            if err != nil {
                log.Println("Error marshaling routing envelope context:", err)
                continue
            }
			if err := targetConn.WriteMessage(websocket.TextMessage, fullMessage); err != nil {
				log.Printf("Failed to forward packet to target '%s': %v", envelope.Target, err)
			}
		}
	}
}

// handleInternalRequest checks if an incoming message is meant for the proxy itself.
// Returns true to drop/ignore the packet, or false if it should be rerouted.
func handleInternalRequest(envelope *WSEnvelope) (drop bool) {
	// Any blank return statements will return true, implying to drop the request
	drop = true

    switch envelope.Event {

    case "get_ping":
		// Update the payload and reroute it to the current target
        playerpingMu.RLock()
        pingData := playerPing
        playerpingMu.RUnlock()

        envelope.Event = "ping_response"
        envelope.Payload, _ = json.Marshal(pingData)

		// Explicitly allow this request to reroute back to the requester
		return false

	case "send_server_packet":
		// Handle incoming server packets requests, using envelope.Target for the player name instead of the 
		session, packetName, packetData, err := extractPacketRequest(envelope)
		if err != nil {
			log.Printf("Server packet extraction failed: %v", err)
			return
		}
		if err := WriteServerJSONPacket(session.ServerConn, packetName, packetData); err != nil {
			log.Printf("Server injection failed for %s: %v", envelope.Target, err)
		}

	case "send_client_packet":
		// Handle incoming client packets requests, using envelope.Target for the player name instead of the 
		session, packetName, packetData, err := extractPacketRequest(envelope)
		if err != nil {
			log.Printf("Client packet extraction failed: %v", err)
			return
		}
		if err := WriteClientJSONPacket(session.ClientConn, packetName, packetData); err != nil {
			log.Printf("Client injection failed for %s: %v", envelope.Target, err)
		}
    }
	// This is not an internal event, so it should not be dropped and should be rerouted
	return false 
}

func SendWebSocketEvent(envelope *WSEnvelope) error {
	// Fetch the target client's WebSocket connection from the registry and process it safely
	wsMu.Lock()
	targetConn, online := wsConns[envelope.Target]
	wsMu.Unlock()

	// If the requested websocket hasn't connected to our proxy yet, fail immediately
	if !online {
		return fmt.Errorf("target client '%s' is offline or not registered", envelope.Target)
	}
	envelope.Timestamp = time.Now().UnixMilli()

	// Transmit the JSON payload specifically down the targeted client socket pipe
	if err := targetConn.WriteJSON(envelope); err != nil {
		return fmt.Errorf("failed to transmit event '%s' to '%s': %w", envelope.Event, envelope.Target, err)
	}
	return nil
}

func RequestWS[T any](target string, event string, data interface{}, timeout time.Duration) (*T, error) {
	// Send a unique ID for this request back to the proxy so that we know if
	requestID := fmt.Sprintf("proxy:%d", atomic.AddUint64(&id_counter, 1))

	// Save this request for the websocket client to fulfill with a reply back
	responseChan := make(chan json.RawMessage, 1)
	pendingRequests.Store(requestID, responseChan)
	defer pendingRequests.Delete(requestID)

	payload, err := json.Marshal(data)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal payload for event '%s': %w", event, err)
	}

	// Send our websocket envelope to the target client with our payload and request ID
	envelope := WSEnvelope{ ID: requestID, Event: event, Target: target, Payload: payload }
	if err := SendWebSocketEvent(&envelope); err != nil {
		return nil, err
	}

	// Wait for our response channel to receive its reply back from target client
	select {
	case rawResponse := <-responseChan:
		var payload T
		if err := json.Unmarshal(rawResponse, &payload); err != nil {
			return nil, fmt.Errorf("failed to parse response type: %w", err)
		}
		// The response is already the payload, not WSEnvelope
		return &payload, nil

	case <-time.After(timeout):
		return nil, fmt.Errorf("request '%s' (%s) timed out", event, requestID)
	}
}

File: discord/src/websocket.js

const WebSocket = require("ws");

/**
 * Model of how the websocket server is going to work. Each client can communicate with one another:
 * 
 * 
 *                       ┌─────────────────────────────────────────────────────────┐
 *                       │                  GO PROXY CENTRAL HUB                   │
 *                       │                                                         │
 *                       │   Active Connection Registry (wsConns Map):             │
 *                       │   {                                                     │
 *                       │     "backend":     *websocket.Conn,                     │
 *                       │     "discord":     *websocket.Conn,                     │
 *                       │     "script_api":  *websocket.Conn                      │
 *                       │   }                                                     │
 *                       │                                                         │
 *                       └────────▲──────────────────▲───────────────────▲─────────┘
 *                                │                  │                   │
 *                                │                  │                   │ Bi-Directional
 *                                │                  │                   │ JSON Envelopes
 *                                ▼                  ▼                   ▼
 *                       ┌──────────────────┐┌──────────────────┐┌──────────────────┐
 *                       │   Node Backend   ││   Discord Bot    ││    Script API    │
 *                       │                  ││                  ││   (BDS Server)   │
 *                       └──────────────────┘└──────────────────┘└──────────────────┘
 * 
 * 
 */

class WebSocketClient {
    constructor() {
        // Used to make sure each outbound request has a unique identifier we can use to map to its promise
        this.id_counter = 0;
        this.ws_url = null;
        this.socket = null;
        this.reconnectTimer = null;
        this.client_name = null;
        this.token = "";

        // Map of request IDs to its promise for requests that expect a response
        this.pendingRequests = new Map();
        this.inboundRequests = new Map();
    }

    /**
     * Start a connection to the provided webscoket url to communicate to our other applications
     * @param {string} name Our arbitrary name of the client that is connecting to the proxy
     * @param {string} base_url The websocket url that we are connecting to (proxy in our case)
     * @param {string} token Auth token to connect to the central hub if defined in the proxy
     * @returns {WebSocketClient}
     */
    connect(name = this.client_name, base_url = this.ws_url, token = this.token) {
        // Save these parameters for later if we try to reconnect without any parameters
        this.client_name = name;
        this.ws_url = base_url;
        this.token = token;
        this.socket = new WebSocket(`${base_url}?client=${name}&token=${token}`);

        // Bind websocket lifecycle event listeners
        this.socket.on("open", () => this.onOpen());
        this.socket.on("message", (message) => this.onMessage(message));
        this.socket.on("close", () => this.onClose());
        this.socket.on("error", (err) => console.error("Socket wire exception:", err.message));
        return this;
    }

    onOpen() {
        // Successfully connected to the central Go Proxy websocket for server handling and communication
        console.log("Successfully connected to the proxy websocket.");
        if (this.reconnectTimer) {
            clearTimeout(this.reconnectTimer);
            this.reconnectTimer = null;
        }
    }

    onClose() {
        // The connection to the proxy websocket was closed. Automatically try to reconnect every 5s
        console.log("Disconnected from the proxy websocket. Reconnecting...");

        // Cancel pending promises so they don't hang forever in the background
        for (const [ id, deferred ] of this.pendingRequests.entries()) {
            deferred.reject(new Error("Connection lost. Request canceled."));
            this.pendingRequests.delete(id);
        }

        // Prevent stacking duplicate timers if multiple close/error events fire
        if (this.reconnectTimer === null) {
            this.reconnectTimer = setTimeout(() => {
                this.reconnectTimer = null;
                this.connect();
            }, 5000);
        }
    }

    /**
     * Send a websocket message to the proxy hub that is then forwarded to the provided target destination
     * @param {string} envelope.event The name of the websocket event so other connections know what is requested 
     * @param {string} envelope.target The destination websocket target that this request is meant for 
     * @param {string} envelope.id The ID of this request. Used for the initial caller to differentiate requests
     * @param {any} envelope.payload The data being sent through the websocket to the proxy hub and then to target
     * @returns {void}
     */
    send(envelope) {
        if (this.isConnected) this.socket.send(JSON.stringify(envelope));
    }

    /**
     * Send an arbitrary websocket request and wait for its response in our onMessage handler event
     * Each request carries its own unique ID so we know which respective promise gets resolved 
     */
    request(event, target, payload = {}, timeout = 5000) {
        return new Promise((resolve, reject) => {
            // Return nothing if the websocket is currently offline or disconnected
            if (!this.isConnected) return resolve(null);

            const requestId = this.client_name + ":" + this.id_counter++;
            this.pendingRequests.set(requestId, { resolve, reject });

            // Timeout safety guard to avoid hangs for requests that don't fulfill quick enough
            setTimeout(() => {
                if (this.pendingRequests.has(requestId)) {
                    this.pendingRequests.delete(requestId);
                    resolve(null);
                }
            }, timeout);

            this.send({ event, target, id: requestId, payload });
        });
    }

    // Used for safely parsing envelopes and return null if they are malformed
    static parse(string) {
        try {
            const result = JSON.parse(string);
            return typeof result === "object" ? result : null;
        } catch { 
            return null;
        }
    }

    // Handle any expected websocket requests that our backend is expected to receive 
    async onMessage(message) {
        const envelope = WebSocketClient.parse(message);
        const { id, event, target, payload } = envelope ?? {};

        // This is a reply to an outbound request we created. Resolve the promise
        if (id && this.pendingRequests.has(id)) {
            const deferred = this.pendingRequests.get(id);
            this.pendingRequests.delete(id);
            return deferred.resolve(payload);
        }

        // This is an inbound request to our script that needs to be fulfilled
        if (this.inboundRequests.has(event))
            return this.inboundRequests.get(event)(envelope);

        console.log(`Unhandled event: ${event}`);
    }

    get isConnected() {
        return this.socket && this.socket.readyState === WebSocket.OPEN;
    }
};

// Connect to the central websocket hub provided a token if needed
const TOKEN = "";
const Client = new WebSocketClient().connect("discord", "ws://127.0.0.1:8080/ws", TOKEN); 

// Expose arbitrary outbound websocket requests to the rest of the program
function requestWS(event, target, payload = {}) {
    return Client.request(event, target, payload);
}

// Sends a request over the websocket with no expected return value back
function sendWS(envelope) {
    return Client.send(envelope);
}

// Allow other files to register their own inbound requests
function registerWsRequest(event, callback) {
    Client.inboundRequests.set(event, callback);
}

module.exports = { requestWS, sendWS, registerWsRequest, WebSocketClient };